fastlane/lib/fastlane/actions/ensure_git_status_clean.rb

```diff
@@ -74,7 +74,7 @@ def self.available_options
                                        type: Boolean),
           FastlaneCore::ConfigItem.new(key: :ignored,
                                        env_name: "FL_ENSURE_GIT_STATUS_CLEAN_IGNORED_FILE",
-                                       description: "The flag whether to ignore file the git status if the repo is dirty",
+                                       description: "The ignored files handling mode if the repo is dirty. The available options are: 'traditional' (default), 'no' and 'matching'",
```

Comment:
Oh, I misinterpreted the git documentation. It states that if you only use `--ignored`, without a value, `traditional` is used - meaning, it's the same as using `--ignored=traditional`.

Reply:
Could you please send me the output of `git status --porcelain --ignored='matching'` too?

Reply:
Sure, here you go:

```
myhost:MY_PROJECT jenkins$ git status --porcelain --ignored='matching'
 M ***/Info.plist
 M Frameworks/***/Info.plist
 M Frameworks/***/ios-arm64/***/Ast
 M Frameworks/***/ios-arm64/***/Info.plist
!! .DS_Store
!! ***/Resources/assets/***/
!! ***/Resources/assets/***/
!! ***/Resources/assets/***/
!! ***/Resources/assets/***/
!! ***/Resources/assets/***/
!! ***/Resources/assets/***/
!! ***/Resources/assets/***/
!! ***/Resources/assets/***.txt
!! ***/Resources/assets/***.xml
!! Gemfile.lock
!! fastlane/README.md
!! fastlane/build/
!! fastlane/report.xml
```

Reply:
> Could you please send me the output of `git status --porcelain --ignored='matching'` too?

Maybe that differs from git version to git version? It would probably a good idea to explicitly pass a default value to `git status --porcelain --ignored=''`

Reply:
> Maybe that differs from git version to git version?

No, I just didn't read it correctly.

> It would probably be a good idea to explicitly pass a default value of `traditional` to `git status --porcelain --ignored=''` if the argument is not set for the fastlane action.

In this case, `git status --porcelain` is used, and I think that it's correct this way, because, if you don't specify it, you'd expect for ignored files NOT to get returned.

I added a test for this case too: https://github.com/fastlane/fastlane/pull/20976/files#diff-72de0e83132788b7354971696dfec7b908ec2009da4808cc9360c1559e870e7eR109-R116.

Technically, the fact that the documentation states that `'no'` is the default value is not quite true, but the effect is the same 😁

Reply:
So I tested this out (thanks for the step-by-step guide) and when you call `bundle exec fastlane run ensure_git_status_clean` the result is as expected the same as calling `git status --porcelain` or `git status --porcelain --ignored='no'`

```
***:*** jenkins$ bundle exec fastlane run ensure_git_status_clean
[20:36:15]: -------------------------------------
[20:36:15]: --- Step: ensure_git_status_clean ---
[20:36:15]: -------------------------------------
[20:36:15]: $ git status --porcelain
[20:36:15]: ▸ M *** /Info.plist
[20:36:15]: ▸ M Frameworks/*** /Info.plist
[20:36:15]: ▸ M Frameworks/*** /ios-arm64/*** /Ast
[20:36:15]: ▸ M Frameworks/*** /ios-arm64/*** /Info.plist
[20:36:15]: ▸ M Gemfile

***:*** jenkins$ git status --porcelain
 M ***/Info.plist
 M Frameworks/***/Info.plist
 M Frameworks/***/ios-arm64/***/Ast
 M Frameworks/***/ios-arm64/***/Info.plist
 M Gemfile

***:*** jenkins$ git status --porcelain --ignored='no'
 M ***/Info.plist
 M Frameworks/***/Info.plist
 M Frameworks/***/ios-arm64/***/Ast
 M Frameworks/***/ios-arm64/***/Info.plist
 M Gemfile
```

But here is the issue: Adding `ignored:'no'` as an argument to the action fails with an error. Passing `ignored:'traditional'` or `ignored:'matching'` works fine and as expected.

```
***:*** jenkins$ bundle exec fastlane run ensure_git_status_clean ignored:'no'
[20:44:11]: You passed invalid parameters to 'ensure_git_status_clean'.
[20:44:11]: Check out the error below and available options by running `fastlane action ensure_git_status_clean`

[!] 'ignored' value must be a String! Found FalseClass instead.

***:*** jenkins$ bundle exec fastlane run ensure_git_status_clean ignored:'traditional'
[20:46:31]: -------------------------------------
[20:46:31]: --- Step: ensure_git_status_clean ---
[20:46:31]: -------------------------------------
[20:46:31]: $ git status --porcelain --ignored='traditional'
[20:46:31]: ▸ M *** /Info.plist
[20:46:31]: ▸ M Frameworks/*** /Info.plist
[20:46:31]: ▸ M Frameworks/*** /ios-arm64/*** /Ast
[20:46:31]: ▸ M Frameworks/*** /ios-arm64/*** /Info.plist
[20:46:31]: ▸ M Gemfile
[20:46:31]: ▸ !! .DS_Store
[20:46:31]: ▸ !! *** /Resources/assets/
[20:46:31]: ▸ !! Gemfile.lock
[20:46:31]: ▸ !! fastlane/README.md
[20:46:31]: ▸ !! fastlane/build/
[20:46:31]: ▸ !! fastlane/report.xml

***:*** jenkins$ bundle exec fastlane run ensure_git_status_clean ignored:'matching'
[20:47:44]: -------------------------------------
[20:47:44]: --- Step: ensure_git_status_clean ---
[20:47:44]: -------------------------------------
[20:47:44]: $ git status --porcelain --ignored='matching'
[20:47:44]: ▸ M ***/Info.plist
[20:47:44]: ▸ M Frameworks/***/Info.plist
[20:47:44]: ▸ M Frameworks/***/ios-arm64/***/Ast
[20:47:44]: ▸ M Frameworks/***/ios-arm64/***/Info.plist
[20:47:44]: ▸ M Gemfile
[20:47:44]: ▸ !! .DS_Store
[20:47:44]: ▸ !! ***/Resources/assets/***/
[20:47:44]: ▸ !! ***/Resources/assets/***/
[20:47:44]: ▸ !! ***/Resources/assets/***/
[20:47:44]: ▸ !! ***/Resources/assets/***/
[20:47:44]: ▸ !! ***/Resources/assets/***/
[20:47:44]: ▸ !! ***/Resources/assets/***/
[20:47:44]: ▸ !! ***/Resources/assets/***/
[20:47:44]: ▸ !! ***/Resources/assets/***.txt
[20:47:44]: ▸ !! ***/Resources/assets/***xml
[20:47:44]: ▸ !! Gemfile.lock
[20:47:44]: ▸ !! fastlane/README.md
[20:47:44]: ▸ !! fastlane/build/
[20:47:44]: ▸ !! fastlane/report.xml
```

Reply:
I found the cause:

https://github.com/fastlane/fastlane/blob/19ab9fad34392514f99b51462e9eec346b197339/fastlane/lib/fastlane/command_line_handler.rb#L43

So the only solution I can think of is to use something else (`'none'`) instead, because that line gets called too early for me to try to prevent it by changing stuff to the `FastlaneCore::ConfigItem` object.

Reply:
I fixed it, so please run `bundle install` (I think) and then `bundle exec fastlane run ensure_git_status_clean ignored:'none'`

Reply:
Looks good!

```
***:*** jenkins$ bundle exec fastlane run ensure_git_status_clean ignored:'none'
[✔] 🚀

[14:10:08]: -------------------------------------
[14:10:08]: --- Step: ensure_git_status_clean ---
[14:10:08]: -------------------------------------
[14:10:08]: $ git status --porcelain --ignored='no'
[14:10:08]: ▸ M ***/Info.plist
[14:10:08]: ▸ M Frameworks/***/Info.plist
[14:10:08]: ▸ M Frameworks/***/ios-arm64/***/Ast
[14:10:08]: ▸ M Frameworks/***/ios-arm64/***/Info.plist
[14:10:08]: ▸ M Gemfile
[14:10:08]: ▸ ?? Gemfile.lock

[!] Git repository is dirty! Please ensure the repo is in a clean state by committing/stashing/discarding all changes first.
```

I also tested the other options (`traditional` and `matching`) and they work as expected.

2 things you might consider adding:

1. Trimming the passed value for `ignored`. If you pass i.e. `ignored:'none '` the command will fail with an exception since `git status` can't handle that.
2. Checking for emptiness after trimming: If you pass i.e. `ignored:''` or `ignored:'    '` both will fail with an exception.

Both of these are just convenience and not a necessity for me to approve this pr. :)

Reply:
> 2 things you might consider adding

Oh, wow, you're amazing! Thanks for catching these too ❤️

Reply:
It looks like other actions don't strip trailing whitespace, so I'll do the same. Meaning, if you specify `'none '`, it will treat it as an invalid value for the `ignored` parameter, and throw an error (which makes sense IMO).

Also, I think that you shouldn't and can't test this validation. Maybe because `FastlaneCore::ConfigItem` are already tested, and it doesn't make sense to test them again in the context of each individual action.

Reply:
You actually did, what I had in mind but was not bold enough to ask for because you did all the work so far. Creating a fixed set of options that can be used as arguments is what I would have done. I am a Swift developer though and it probably would have taken me days to figure this out. 🤣 

Great work! Thanks.